p2p/transport/tcp/conn/connection.go

```diff
@@ -269,147 +233,181 @@ func (c *MConnection) stopServices() (alreadyStopped bool) {
-	// recvRoutine may write to it after we've stopped.
-	// Though it doesn't need to get closed at all,
-	// we close it @ recvRoutine.
-
-	// c.Stop()
+	return c.conn.Close()
 }
 
-// OnStop implements BaseService.
-func (c *MConnection) OnStop() {
-	if c.stopServices() {
-		return
-	}
+func (c *MConnection) ConnState() (state transport.ConnState) {
+	state.ConnectedFor = time.Since(c.created)
+	state.SendRateLimiterDelay = c.sendMonitor.Status().SleepTime
+	state.RecvRateLimiterDelay = c.recvMonitor.Status().SleepTime
+	state.StreamStates = make(map[byte]transport.StreamState)
 
-	c.conn.Close()
+	for streamID, channel := range c.channelsIdx {
+		state.StreamStates[streamID] = transport.StreamState{
+			SendQueueSize:     channel.loadSendQueueSize(),
+			SendQueueCapacity: cap(channel.sendQueue),
+		}
+	}
 
-	// We can't close pong safely here because
-	// recvRoutine may write to it after we've stopped.
-	// Though it doesn't need to get closed at all,
-	// we close it @ recvRoutine.
+	return state
 }
 
 func (c *MConnection) String() string {
 	return fmt.Sprintf("MConn{%v}", c.conn.RemoteAddr())
 }
 
-func (c *MConnection) flush() {
-	c.Logger.Debug("Flush", "conn", c)
-	err := c.bufConnWriter.Flush()
-	if err != nil {
-		c.Logger.Debug("MConnection flush failed", "err", err)
-	}
+func (c *MConnection) flush() error {
+	return c.bufConnWriter.Flush()
 }
 
 // Catch panics, usually caused by remote disconnects.
 func (c *MConnection) _recover() {
 	if r := recover(); r != nil {
 		c.Logger.Error("MConnection panicked", "err", r, "stack", string(debug.Stack()))
-		c.stopForError(fmt.Errorf("recovered from panic: %v", r))
-	}
-}
-
-func (c *MConnection) stopForError(r any) {
-	if err := c.Stop(); err != nil {
-		c.Logger.Error("Error stopping connection", "err", err)
-	}
-	if atomic.CompareAndSwapUint32(&c.errored, 0, 1) {
-		if c.onError != nil {
-			c.onError(r)
-		}
+		c.Close(fmt.Sprintf("recovered from panic: %v", r))
 	}
 }
 
-// Queues a message to be sent to channel.
-func (c *MConnection) Send(chID byte, msgBytes []byte) bool {
+// thread-safe.
+func (c *MConnection) sendBytes(chID byte, msgBytes []byte, blocking bool) error {
 	if !c.IsRunning() {
-		return false
+		return nil
 	}
 
-	c.Logger.Debug("Send", "channel", chID, "conn", c, "msgBytes", log.NewLazySprintf("%X", msgBytes))
+	// c.Logger.Debug("Send",
+	// 	"streamID", chID,
+	// 	"msgBytes", log.NewLazySprintf("%X", msgBytes),
+	// 	"timeout", timeout)
 
-	// Send message to channel.
 	channel, ok := c.channelsIdx[chID]
 	if !ok {
-		c.Logger.Error(fmt.Sprintf("Cannot send bytes, unknown channel %X", chID))
-		return false
-	}
-
-	success := channel.sendBytes(msgBytes)
-	if success {
-		// Wake up sendRoutine if necessary
-		select {
-		case c.send <- struct{}{}:
-		default:
-		}
-	} else {
-		c.Logger.Debug("Send failed", "channel", chID, "conn", c, "msgBytes", log.NewLazySprintf("%X", msgBytes))
+		panic(fmt.Sprintf("Unknown channel %X. Forgot to register?", chID))
```

Comment:
Are we sure we want to panic and crash the node?  Can we have a situation where the channel was for some reason not yet registered but there is something being sent to it ?

Reply:
> Are we sure we want to panic and crash the node? Can we have a situation where the channel was for some reason not yet registered but there is something being sent to it ?

Yes. We have discussed this with Daniel in [the https://github.com/cometbft/cometbft/pull/4307](https://github.com/cometbft/cometbft/pull/4307#discussion_r1858988094). We (devs) should make sure all streams are registered before starting `MConnection`.